ios-base/Networking/APIService.swift

```diff
@@ -81,84 +84,70 @@ open class BaseApiService<T>: APIService where T: TargetType {
 
   /**
    Makes a request to the provided target and tries to decode its response
-   using the provided keyPath and return type and returning it on the onSuccess callback.
+   using the provided keyPath and return type and returning it as an Observable.
    - Parameters:
    - target: The TargetType used to make the request.
    - keyPath: The keypath used to decode from JSON (if passed nil, it will try to decode from the root).
    */
-  open func request<T>(for target: ProviderType,
-                       at keyPath: String? = nil,
-                       onSuccess: @escaping (_ result: T, _ response: Response) -> Void,
-                       onFailure: ((Error, Response?) -> Void)? = nil) where T: Codable {
-    provider.request(target, completion: { [weak self] result in
-      guard let self = self else { return }
-
-      switch result {
-      case let .success(moyaResponse):
-        self.handleSuccess(with: moyaResponse, at: keyPath, onSuccess: onSuccess, onFailure: onFailure)
-      case let .failure(error):
-        self.handleError(with: error, onFailure)
+  open func request<T>(for target: ProviderType, at keyPath: String? = nil) -> Observable<(T, Response)> where T: Codable {
+    return provider.rx.request(target)
+      .filterSuccessfulStatusCodes()
+      .flatMap { [weak self] response in
+        let decodedValue = try response.map(T.self,
+                                atKeyPath: keyPath,
+                                using: self?.jsonDecoder ?? JSONDecoder(),
+                                failsOnEmptyData: true)
+        return .just((decodedValue, response))
+      }
+      .asObservable()
+      .catchError { [weak self] error in
+        guard let self = self else {
+          return Observable.error(error)
+        }
+        return Observable.error(self.handleError(with: error))
       }
-    })
   }
 
   /**
-   Makes a request to the provided target
+   Makes a request to the provided target and returns as an Obserbable
+   - Parameters:
+   - target: The TargetType used to make the request.
    */
-  open func request(for target: ProviderType,
-                    onSuccess: @escaping (_ response: Response) -> Void,
-                    onFailure: ((Error, Response?) -> Void)? = nil) {
-    provider.request(target, completion: { [weak self] result in
-      guard let self = self else { return }
-
-      switch result {
-      case let .success(moyaResponse):
-        onSuccess(moyaResponse)
-      case let .failure(error):
-        self.handleError(with: error, onFailure)
+  open func request(for target: ProviderType) -> Observable<Response> {
+    return provider.rx.request(target)
+      .filterSuccessfulStatusCodes()
+      .asObservable()
+      .catchError { [weak self] error in
+        guard let self = self else {
+          return Observable.error(error)
+        }
+        return Observable.error(self.handleError(with: error))
```

Comment:
does `Observable.error` accepts optionals ? in that case I'd just return 
`Observable.error(self?.handleError(with: error))` and lose the guard

Reply:
It doesn't, since Observable secuences can either finish by completing or error out, the error case must be a concrete error, not nil.